Many symintifications

albanD opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #87722
* #87294
* __->__ #87604

Adds
expand_inplace
conv conv_double_backward
convolution
adaptive_avg_pool2d_symint
_embedding_bag_backward_symint
cudnn_grid_sampler
cuda 32 bit indexing
nll_loss / nll_loss_2d
tensor split
pooling same mode
cudnn_is_acceptable
storage nbytes

albanD committed:
Many symintifications

[ghstack-poisoned]

albanD commented:
Will wait for CI for all the xfail update and make sure everything is tested.

bdhirsh commented on aten/src/ATen/native/NonSymbolicBC.h:
```diff
@@ -22,4 +22,5 @@ TORCH_API at::Tensor _embedding_bag_sparse_backward(const at::Tensor & grad, con
 TORCH_API at::Tensor value_selecting_reduction_backward(const at::Tensor & grad, int64_t dim, const at::Tensor & indices, at::IntArrayRef sizes, bool keepdim);
 TORCH_API at::Tensor trace_backward(const at::Tensor & grad, at::IntArrayRef sizes);
 TORCH_API at::Tensor index_select_backward(const at::Tensor & grad, at::IntArrayRef self_sizes, int64_t dim, const at::Tensor & index);
+TORCH_API std::vector<Tensor> tensor_split(const Tensor& self, IntArrayRef indices, int64_t dim);
```
This has a small risk of internal breakage, right?

Although I just searched internal for `at::native::tensor_split`, and the only place that shows up is in static runtime (which is part of the OSS codebase). So.. I think this should be fine

albanD replied:
Ok, let's take the risk on this one. Thanks for checking!

bdhirsh commented on aten/src/ATen/native/Pool.h:
```diff
@@ -67,24 +67,34 @@ static inline T pooling_output_shape(
     auto wiggle_room = inputSize % stride - 1;
     if (wiggle_room > 0) {
-      --total_padding;
+      total_padding = total_padding - 1;
```
Do you think it's worth implementing `operator--` as part of merging to master? (I'll leave it up to you)

ezyang replied:
seems worthwhile. But I'll take a rain check if someone agrees to do ALL of the operator overloads, methodically

albanD replied:
Ok, leaving all the operators for a follow up then.

bdhirsh commented on aten/src/ATen/native/TensorShape.cpp:
```diff
@@ -756,21 +759,30 @@ std::vector<Tensor> tensor_split(const Tensor& self, int64_t sections, int64_t d
     start_idx = end_idx;
   }
-  splits[num_indices] = at::slice(self, dim_, start_idx, self.size(dim_));
+  splits[num_indices] = at::symint::slice<T>(self, dim_, start_idx, self.size(dim_));
```
hmmm don't we need to symintify that `self.size(dim_)`?

I guess we could add a template for `at::symint::size()`, and make it:
```
at::symint::size<T>(self)`
```
wdyt?

albanD replied:
Yes I can add that.

bdhirsh commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -1815,7 +1821,7 @@
   variants: function, method
   dispatch:
-    CompositeExplicitAutogradNonFunctional: diag_embed
+    CompositeExplicitAutograd: diag_embed
```
hmmm, why?

`CompositExplicitAutogradNonFunctional` is a bit funny today - it's basically only used for providing better error messages for XLA/LTC. If we have a functional op that calls a view op, that is `CompositeExplicitAutograd`, then XLA and LTC can't handle it directly so we don't want to register it to them. Not clear to me why we had to remove it for `diag_embed` here though.

albanD replied:
Not sure tbh, I haven't reviewed this. Reverting.

bdhirsh commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -7698,10 +7704,21 @@
+    CUDA: diag_cuda_out
+    MPS: diag_mps_out
 
 - func: diag(Tensor self, int diagonal=0) -> Tensor
```
Is there a comment explaining the reasoning for the change to `diag`?

It looks like before, `diag` was a `CompositeImplicitAutograd` op that called `diag_embed`, and now we have a brand new backward op for it.

albanD replied:
This is a mistake. The diag improvements were added to master indepently and never added to the branch.
I'll make sure not to revert https://github.com/pytorch/pytorch/pull/87180 by mistake :p

bdhirsh commented on aten/src/ATen/native/TensorShape.cpp:
```diff
@@ -3666,16 +3721,8 @@ at::Tensor& _sparse_broadcast_to_copy_out(const at::Tensor & self, at::IntArrayR
-  TORCH_CHECK(
-      canCast(result.scalar_type(), out.scalar_type()),
-      "diagonal_copy: result type ", result.scalar_type(), " can't be cast to the desired out= type ", out.scalar_type());
-  out.copy_(result);
```
This seems wrong - we're removing error checking?

albanD committed:
Update on "Many symintifications"


Adds
expand_inplace
conv conv_double_backward
convolution
adaptive_avg_pool2d_symint
_embedding_bag_backward_symint
cudnn_grid_sampler
cuda 32 bit indexing
nll_loss / nll_loss_2d
tensor split
pooling same mode
cudnn_is_acceptable
storage nbytes


[ghstack-poisoned]

albanD commented on torchgen/api/cpp.py:
```diff
@@ -342,11 +342,13 @@ def default_expr(d: str, t: Type) -> str:
         if d.startswith("[") and d.endswith("]"):
             return "{" + d[1:-1] + "}"
+        elif symint and d.isdigit() and str(t.elem) == "SymInt":
+            return f"c10::SymInt({d})"
```
It is actually used for `SymInt[3] padding=0` so it is for the list case.

albanD committed:
Update on "Many symintifications"


Adds
expand_inplace
conv conv_double_backward
convolution
adaptive_avg_pool2d_symint
_embedding_bag_backward_symint
cudnn_grid_sampler
cuda 32 bit indexing
nll_loss / nll_loss_2d
tensor split
pooling same mode
cudnn_is_acceptable
storage nbytes


[ghstack-poisoned]

albanD committed:
Update on "Many symintifications"


Adds
expand_inplace
conv conv_double_backward
convolution
adaptive_avg_pool2d_symint
_embedding_bag_backward_symint
cudnn_grid_sampler
cuda 32 bit indexing
nll_loss / nll_loss_2d
tensor split
pooling same mode
cudnn_is_acceptable
storage nbytes


[ghstack-poisoned]

albanD committed:
Update on "Many symintifications"


Adds
expand_inplace
conv conv_double_backward
convolution
adaptive_avg_pool2d_symint
_embedding_bag_backward_symint
cudnn_grid_sampler
cuda 32 bit indexing
nll_loss / nll_loss_2d
tensor split
pooling same mode
cudnn_is_acceptable
storage nbytes


[ghstack-poisoned]

ezyang commented on aten/src/ATen/native/TensorShape.cpp:
```diff
@@ -3390,29 +3402,72 @@ Tensor unfold(const Tensor& self, int64_t d, int64_t size, int64_t step) {
-  if (ndim == 1) {
-    return at::diag_embed(self, offset);
+template <typename scalar_t>
+void apply_diag(Tensor& result, const Tensor& self, int64_t dimension) {
```
The diff here is weirdly big and makes me nervous

ezyang commented:
Everything here looks good except diag, which is weirdly complicated for what I expect

albanD commented:
>  Everything here looks good except diag, which is weirdly complicated for what I expect

Yes the diag change is a mistake due to a land race with master. Cleaning this up now.

albanD committed:
Update on "Many symintifications"


Adds
expand_inplace
conv conv_double_backward
convolution
adaptive_avg_pool2d_symint
_embedding_bag_backward_symint
cudnn_grid_sampler
cuda 32 bit indexing
nll_loss / nll_loss_2d
tensor split
pooling same mode
cudnn_is_acceptable
storage nbytes


[ghstack-poisoned]

albanD committed:
Update on "Many symintifications"


Adds
expand_inplace
conv conv_double_backward
convolution
adaptive_avg_pool2d_symint
_embedding_bag_backward_symint
cudnn_grid_sampler
cuda 32 bit indexing
nll_loss / nll_loss_2d
tensor split
pooling same mode
cudnn_is_acceptable
storage nbytes


[ghstack-poisoned]

albanD committed:
Update on "Many symintifications"


Adds
expand_inplace
conv conv_double_backward
convolution
adaptive_avg_pool2d_symint
_embedding_bag_backward_symint
cudnn_grid_sampler
cuda 32 bit indexing
nll_loss / nll_loss_2d
tensor split
pooling same mode
cudnn_is_acceptable
storage nbytes


[ghstack-poisoned]

albanD committed:
Update on "Many symintifications"


Adds
expand_inplace
conv conv_double_backward
convolution
adaptive_avg_pool2d_symint
_embedding_bag_backward_symint
cudnn_grid_sampler
cuda 32 bit indexing
nll_loss / nll_loss_2d
tensor split
pooling same mode
cudnn_is_acceptable
storage nbytes


[ghstack-poisoned]

albanD committed:
Update on "Many symintifications"


Adds
expand_inplace
conv conv_double_backward
convolution
adaptive_avg_pool2d_symint
_embedding_bag_backward_symint
cudnn_grid_sampler
cuda 32 bit indexing
nll_loss / nll_loss_2d
tensor split
pooling same mode
cudnn_is_acceptable
storage nbytes


[ghstack-poisoned]

weiwangmeta commented:
@pytorchbot revert -m "breaking internal builds" -c ghfirst
